pkg/server/cloud/gce/firewall.go

```diff
@@ -0,0 +1,304 @@
+		log.Printf("error converting port %s: %v", portStr, err)
+	}
+	if endPortStr != "" {
+		endPort, err := strconv.Atoi(endPortStr)
+		if err != nil {
+			log.Printf("error converting ending port of range %s: %v", endPortStr, err)
+		}
+		if endPort > port {
+			size = endPort - port + 1
+		}
+	}
+	return port, size
+}
+
+func allowedRulesToPorts(fwa []*compute.FirewallAllowed) []cloud.InstancePort {
+	instancePorts := make([]cloud.InstancePort, 0, len(fwa))
+	for _, fw := range fwa {
+		if fw == nil {
+			continue
+		}
+		proto := gceProtocolToKipProtocol(fw.IPProtocol)
+		if proto == "" {
+			klog.Warningln("Unknown protocol in firewall rule", fw.IPProtocol)
+			continue
+		}
+		for _, port := range fw.Ports {
+			port, size := portStringToInstancePort(port)
+			instancePorts = append(instancePorts, cloud.InstancePort{
+				Protocol:      proto,
+				Port:          port,
+				PortRangeSize: size,
+			})
+		}
+	}
+	return instancePorts
+}
+
+func firewallToSecurityGroup(fw *compute.Firewall) *cloud.SecurityGroup {
+	return &cloud.SecurityGroup{
+		Name:         fw.Name,
+		ID:           fw.Name,
+		Ports:        allowedRulesToPorts(fw.Allowed),
+		SourceRanges: fw.SourceRanges,
+	}
+}
+
+func (c *gceClient) toFirewallRule(sgName string, ports []cloud.InstancePort, sourceRanges []string) *compute.Firewall {
+	return &compute.Firewall{
+		Allowed:      portsToAllowedRules(ports),
+		Description:  c.getRuleDescription(),
+		Direction:    "INGRESS",
+		Name:         sgName,
+		Network:      c.getNetworkURL(),
+		SourceRanges: sourceRanges,
+		TargetTags:   []string{CreateKipCellNetworkTag(c.controllerID)},
+	}
+}
+
+func (c *gceClient) EnsureMilpaSecurityGroups(extraCIDRs, extraGroupIDs []string) error {
+	milpaPorts := []cloud.InstancePort{
+		{
+			Protocol:      api.ProtocolTCP,
+			Port:          cloud.RestAPIPort,
+			PortRangeSize: 1,
+		},
+		{
+			Protocol:      api.ProtocolTCP,
+			Port:          1,
+			PortRangeSize: math.MaxUint16,
+		},
+		{
+			Protocol:      api.ProtocolUDP,
+			Port:          1,
+			PortRangeSize: math.MaxUint16,
+		},
+		{
+			Protocol:      api.ProtocolICMP,
+			Port:          -1,
+			PortRangeSize: 1,
+		},
+	}
+	cidrs := make([]string, len(c.vpcCIDRs))
+	copy(cidrs, c.vpcCIDRs)
+	for _, cidr := range extraCIDRs {
+		if cidr != "" {
+			cidrs = append(cidrs, cidr)
+		}
+	}
+
+	apiGroupName := CreateKipCellNetworkTag(c.controllerID)
+	apiGroup, err := c.EnsureSecurityGroup(apiGroupName, milpaPorts, cidrs)
+	if err != nil {
+		return util.WrapError(err, "could not setup Kip API cloud firewall rules")
+	}
+	ids := append(extraGroupIDs, apiGroup.ID)
+	c.SetBootSecurityGroupIDs(ids)
+	return nil
+}
+
+func (c *gceClient) FindSecurityGroup(sgName string) (*cloud.SecurityGroup, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+	defer cancel()
+	resp, err := c.service.Firewalls.Get(c.projectID, sgName).Context(ctx).Do()
+	if err != nil {
+		if isNotFoundError(err) {
+			return nil, nil
```

Comment:
commit aef59c3 fixes this

Reply:
The callers of this code assume that FindSecurityGroup will not return an error if the security group is not found.  Is it an error for a find operation to not find anything?  In AWS it is not an error, you just get back a list of security groups with len == 0.  We've gone with that convention in GCE as well.  I went with that behavior since its actually easier to deal with that vs. checking error types.

Personally, I'd like to keep things as they were but the question is where we want consistency. Long ago, FindSecurityGroup was in the CloudClient interface so we'd want to stay consistent with the implementation in AWS.  It's no longer in the interface...  If you'd really like for this to return an error and you think it should behave differently than AWS when a security group isn't found, please update callers of this as well to handle the error appropriately.  By returning an error, callers of this fuction now break.

Reply:
Understood I actually just experienced this now. I have since reverted this issue, thank you for pointing this out @justnoise